### PR TITLE
executor: use helper.GetRegionsTableInfo to get region schema for tidb_hot_regions_history

### PR DESCRIPTION
**Pull request by IcePigZDB**
Signed-off-by: IcePigZDB <icepigzdb@gmail.com>
### What problem does this PR solve?

Issue Number:  #30518

Problem Summary:
`TIDB_HOT_REGIONS_HISTORY` need to parse shcema info on TiDB according to the key range of every Region. However, after v5.0 [`enable-cross-table-merge` is default to true](https://docs.pingcap.com/tidb/stable/release-5.0.0/#configuration-file-parameters), a region may contain multi tables or indics,  Previous implemention #27375 `tidb/store/helper/Helper.FindTableIndexOfRegion` iterate over all tables(first table record and then index ) to check whether it is contained in the region and stop iterate  when it find the first table or index.  This may result in incomplete result,  so use `GetTablesInfoWithKeyRange` for `TIDB_HOT_REGIONS_HISTORY`.
* Use FilterMemDBs to filter Memory tables who have no corresponding regions.
* Use `GetRegionsTableInfo`  instead of `FindTableIndexOfRegion`.
* Extract `GetTablesInfoWithKeyRange` from `GetRegionsTableInfo` to avoid repeat call.
* Set metrics'(`HOT_DEGREE`, `HOT_DEGREE`, `KEY_RATE`, `QUERY_RATE`) default value instead of set NULL when they are 0.

### Change Preview
* Compare the parse result with old and new implmention.
	* old result 1
	* old result 2
	* new result
	* selected region's schemas in tikv region status 

I list top 10 region in time order for three query and check region's schema with `tikv_region_status`, new implemention is consistent with `tikv_region_status`. Complete result is [here](https://github.com/IcePigZDB/discussion-refs/tree/master/tidb_hot_regions_history).

Compared with the old Schema parsing method, the new method also parses the index of the region, it will parse other tables too if there are. Due to the allSchemas' order may different each query, it leads to inconsistent query results.

* old result 1
```sql
mysql> select * from information_schema.tidb_hot_regions_history where update_time>='2015-04-10 10:10:10' and update_time<='2022-05-10 10:10:10';
+---------------------+--------------------+---------------+---------------------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
| UPDATE_TIME         | DB_NAME            | TABLE_NAME    | TABLE_ID            | INDEX_NAME | INDEX_ID | REGION_ID | STORE_ID | PEER_ID | IS_LEARNER | IS_LEADER | TYPE | HOT_DEGREE | HOT_DEGREE | KEY_RATE | QUERY_RATE |
+---------------------+--------------------+---------------+---------------------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |        83 |        1 |     129 |          0 |         1 | READ |         58 |      82752 |     NULL |        404 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |        85 |        1 |     166 |          0 |         1 | READ |         58 |      86136 |     NULL |        412 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |        89 |        1 |      98 |          0 |         1 | READ |         58 |      85657 |     NULL |        410 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |        95 |        1 |     113 |          0 |         1 | READ |         58 |      85118 |     NULL |        407 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |       100 |        1 |     170 |          0 |         1 | READ |         58 |      86316 |     NULL |        413 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |       103 |        1 |     124 |          0 |         1 | READ |         58 |      85624 |     NULL |        410 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |       110 |        1 |     173 |          0 |         1 | READ |         58 |      85934 |     NULL |        411 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |       122 |        1 |     162 |          0 |         1 | READ |         58 |      86065 |     NULL |        412 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |       127 |        1 |     168 |          0 |         1 | READ |         58 |      85529 |     NULL |        409 |
| 2021-12-29 01:45:29 | PERFORMANCE_SCHEMA | GLOBAL_STATUS | 4611686018427397905 | NULL       |     NULL |       140 |        1 |     177 |          0 |         1 | READ |         58 |      85479 |     NULL |        409 |
+---------------------+--------------------+---------------+---------------------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
```
* old result 2
```sql
mysql> select * from information_schema.tidb_hot_regions_history where update_time>='2015-04-10 10:10:10' and update_time<='2022-05-10 10:10:10';
+---------------------+---------+------------+----------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
| UPDATE_TIME         | DB_NAME | TABLE_NAME | TABLE_ID | INDEX_NAME | INDEX_ID | REGION_ID | STORE_ID | PEER_ID | IS_LEARNER | IS_LEADER | TYPE | HOT_DEGREE | FLOW_BYTES | KEY_RATE | QUERY_RATE |
+---------------------+---------+------------+----------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |        83 |        1 |     129 |          0 |         1 | READ |         58 |      82752 |     NULL |        404 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |        85 |        1 |     166 |          0 |         1 | READ |         58 |      86136 |     NULL |        412 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |        89 |        1 |      98 |          0 |         1 | READ |         58 |      85657 |     NULL |        410 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |        95 |        1 |     113 |          0 |         1 | READ |         58 |      85118 |     NULL |        407 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |       100 |        1 |     170 |          0 |         1 | READ |         58 |      86316 |     NULL |        413 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |       103 |        1 |     124 |          0 |         1 | READ |         58 |      85624 |     NULL |        410 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |       110 |        1 |     173 |          0 |         1 | READ |         58 |      85934 |     NULL |        411 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |       122 |        1 |     162 |          0 |         1 | READ |         58 |      86065 |     NULL |        412 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |       127 |        1 |     168 |          0 |         1 | READ |         58 |      85529 |     NULL |        409 |
| 2021-12-29 01:45:29 | TEST    | T1000      |      157 | NULL       |     NULL |       140 |        1 |     177 |          0 |         1 | READ |         58 |      85479 |     NULL |        409 |
+---------------------+---------+------------+----------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
```

* new result
```sql
mysql> select * from information_schema.tidb_hot_regions_history where update_time>='2015-04-10 10:10:10' and update_time<='2022-05-10 10:10:10';
+---------------------+---------+------------+----------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
| UPDATE_TIME         | DB_NAME | TABLE_NAME | TABLE_ID | INDEX_NAME | INDEX_ID | REGION_ID | STORE_ID | PEER_ID | IS_LEARNER | IS_LEADER | TYPE | HOT_DEGREE | FLOW_BYTES | KEY_RATE | QUERY_RATE |
+---------------------+---------+------------+----------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
| 2021-12-29 01:45:29 | SBTEST  | SBTEST5    |       59 | K_5        |        1 |        83 |        1 |     129 |          0 |         1 | READ |         58 |      82752 |        0 |        404 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST5    |       59 | NULL       |     NULL |        83 |        1 |     129 |          0 |         1 | READ |         58 |      82752 |        0 |        404 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST3    |       62 | K_3        |        1 |        85 |        1 |     166 |          0 |         1 | READ |         58 |      86136 |        0 |        412 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST3    |       62 | NULL       |     NULL |        85 |        1 |     166 |          0 |         1 | READ |         58 |      86136 |        0 |        412 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST8    |       63 | K_8        |        1 |        89 |        1 |      98 |          0 |         1 | READ |         58 |      85657 |        0 |        410 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST8    |       63 | NULL       |     NULL |        89 |        1 |      98 |          0 |         1 | READ |         58 |      85657 |        0 |        410 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST1    |       73 | K_1        |        1 |        95 |        1 |     113 |          0 |         1 | READ |         58 |      85118 |        0 |        407 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST1    |       73 | NULL       |     NULL |        95 |        1 |     113 |          0 |         1 | READ |         58 |      85118 |        0 |        407 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST2    |       75 | K_2        |        1 |       100 |        1 |     170 |          0 |         1 | READ |         58 |      86316 |        0 |        413 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST2    |       75 | NULL       |     NULL |       100 |        1 |     170 |          0 |         1 | READ |         58 |      86316 |        0 |        413 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST13   |       83 | K_13       |        1 |       103 |        1 |     124 |          0 |         1 | READ |         58 |      85624 |        0 |        410 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST13   |       83 | NULL       |     NULL |       103 |        1 |     124 |          0 |         1 | READ |         58 |      85624 |        0 |        410 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST11   |       92 | K_11       |        1 |       110 |        1 |     173 |          0 |         1 | READ |         58 |      85934 |        0 |        411 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST11   |       92 | NULL       |     NULL |       110 |        1 |     173 |          0 |         1 | READ |         58 |      85934 |        0 |        411 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST10   |      104 | K_10       |        1 |       122 |        1 |     162 |          0 |         1 | READ |         58 |      86065 |        0 |        412 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST10   |      104 | NULL       |     NULL |       122 |        1 |     162 |          0 |         1 | READ |         58 |      86065 |        0 |        412 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST24   |      110 | K_24       |        1 |       127 |        1 |     168 |          0 |         1 | READ |         58 |      85529 |        0 |        409 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST24   |      110 | NULL       |     NULL |       127 |        1 |     168 |          0 |         1 | READ |         58 |      85529 |        0 |        409 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST23   |      122 | K_23       |        1 |       140 |        1 |     177 |          0 |         1 | READ |         58 |      85479 |        0 |        409 |
| 2021-12-29 01:45:29 | SBTEST  | SBTEST23   |      122 | NULL       |     NULL |       140 |        1 |     177 |          0 |         1 | READ |         58 |      85479 |        0 |        409 |
+---------------------+---------+------------+----------+------------+----------+-----------+----------+---------+------------+-----------+------+------------+------------+----------+------------+
```
* selected region's schemas in tikv region status 
```sql
mysql> select * from information_schema.tikv_region_status where region_id in (83,85,89,95,100,103,110,122,127,140) order by region_id;
+-----------+--------------------------------------+--------------------------------------+----------+---------+------------+----------+----------+------------+----------------+---------------+---------------+------------+------------------+------------------+-------------------------+---------------------------+
| REGION_ID | START_KEY                            | END_KEY                              | TABLE_ID | DB_NAME | TABLE_NAME | IS_INDEX | INDEX_ID | INDEX_NAME | EPOCH_CONF_VER | EPOCH_VERSION | WRITTEN_BYTES | READ_BYTES | APPROXIMATE_SIZE | APPROXIMATE_KEYS | REPLICATIONSTATUS_STATE | REPLICATIONSTATUS_STATEID |
+-----------+--------------------------------------+--------------------------------------+----------+---------+------------+----------+----------+------------+----------------+---------------+---------------+------------+------------------+------------------+-------------------------+---------------------------+
|        83 | 7480000000000000FF3B00000000000000F8 | 7480000000000000FF3D00000000000000F8 |       59 | sbtest  | sbtest5    |        1 |        1 | k_5        |             13 |            29 |            30 |          0 |               29 |           205024 | NULL                    |                      NULL |
|        83 | 7480000000000000FF3B00000000000000F8 | 7480000000000000FF3D00000000000000F8 |       59 | sbtest  | sbtest5    |        0 |     NULL | NULL       |             13 |            29 |            30 |          0 |               29 |           205024 | NULL                    |                      NULL |
|        85 | 7480000000000000FF3E00000000000000F8 | 7480000000000000FF3F00000000000000F8 |       62 | sbtest  | sbtest3    |        0 |     NULL | NULL       |             13 |            31 |            30 |          0 |               30 |           201374 | NULL                    |                      NULL |
|        85 | 7480000000000000FF3E00000000000000F8 | 7480000000000000FF3F00000000000000F8 |       62 | sbtest  | sbtest3    |        1 |        1 | k_3        |             13 |            31 |            30 |          0 |               30 |           201374 | NULL                    |                      NULL |
|        89 | 7480000000000000FF3F00000000000000F8 | 7480000000000000FF4400000000000000F8 |       63 | sbtest  | sbtest8    |        0 |     NULL | NULL       |             13 |            31 |            30 |          0 |               29 |           200376 | NULL                    |                      NULL |
|        89 | 7480000000000000FF3F00000000000000F8 | 7480000000000000FF4400000000000000F8 |       63 | sbtest  | sbtest8    |        1 |        1 | k_8        |             13 |            31 |            30 |          0 |               29 |           200376 | NULL                    |                      NULL |
|        95 | 7480000000000000FF4900000000000000F8 | 7480000000000000FF4B00000000000000F8 |       73 | sbtest  | sbtest1    |        0 |     NULL | NULL       |             13 |            34 |            30 |          0 |               30 |           203195 | NULL                    |                      NULL |
|        95 | 7480000000000000FF4900000000000000F8 | 7480000000000000FF4B00000000000000F8 |       73 | sbtest  | sbtest1    |        1 |        1 | k_1        |             13 |            34 |            30 |          0 |               30 |           203195 | NULL                    |                      NULL |
|       100 | 7480000000000000FF4B00000000000000F8 | 7480000000000000FF5300000000000000F8 |       75 | sbtest  | sbtest2    |        1 |        1 | k_2        |             19 |            35 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       100 | 7480000000000000FF4B00000000000000F8 | 7480000000000000FF5300000000000000F8 |       75 | sbtest  | sbtest2    |        0 |     NULL | NULL       |             19 |            35 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       103 | 7480000000000000FF5300000000000000F8 | 7480000000000000FF5600000000000000F8 |       83 | sbtest  | sbtest13   |        1 |        1 | k_13       |             19 |            36 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       103 | 7480000000000000FF5300000000000000F8 | 7480000000000000FF5600000000000000F8 |       83 | sbtest  | sbtest13   |        0 |     NULL | NULL       |             19 |            36 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       110 | 7480000000000000FF5C00000000000000F8 | 7480000000000000FF5F00000000000000F8 |       92 | sbtest  | sbtest11   |        1 |        1 | k_11       |             19 |            39 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       110 | 7480000000000000FF5C00000000000000F8 | 7480000000000000FF5F00000000000000F8 |       92 | sbtest  | sbtest11   |        0 |     NULL | NULL       |             19 |            39 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       122 | 7480000000000000FF6800000000000000F8 | 7480000000000000FF6B00000000000000F8 |      104 | sbtest  | sbtest10   |        1 |        1 | k_10       |             19 |            43 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       122 | 7480000000000000FF6800000000000000F8 | 7480000000000000FF6B00000000000000F8 |      104 | sbtest  | sbtest10   |        0 |     NULL | NULL       |             19 |            43 |            30 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       127 | 7480000000000000FF6E00000000000000F8 | 7480000000000000FF7100000000000000F8 |      110 | sbtest  | sbtest24   |        1 |        1 | k_24       |             19 |            45 |             0 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       127 | 7480000000000000FF6E00000000000000F8 | 7480000000000000FF7100000000000000F8 |      110 | sbtest  | sbtest24   |        0 |     NULL | NULL       |             19 |            45 |             0 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       140 | 7480000000000000FF7A00000000000000F8 | 7480000000000000FF7D00000000000000F8 |      122 | sbtest  | sbtest23   |        1 |        1 | k_23       |             19 |            49 |             0 |          0 |               29 |           200000 | NULL                    |                      NULL |
|       140 | 7480000000000000FF7A00000000000000F8 | 7480000000000000FF7D00000000000000F8 |      122 | sbtest  | sbtest23   |        0 |     NULL | NULL       |             19 |            49 |             0 |          0 |               29 |           200000 | NULL                    |                      NULL |
+-----------+--------------------------------------+--------------------------------------+----------+---------+------------+----------+----------+------------+----------------+---------------+---------------+------------+------------------+------------------+-------------------------+---------------------------+
20 rows in set (9.88 sec)
```
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by IcePigZDB** [comment]
您好，我已经收到了你的来信。我会及时回复您的。祝工作顺利，身体健康。

File: executor/memtable_reader.go [hunk 1]
```diff
@@ -926,74 +926,77 @@ func (e *hotRegionsHistoryRetriver) retrieve(ctx context.Context, sctx sessionct
...
-	}
-	if hisHotRegion.IsLeader {
-		row[10].SetInt64(1)
-	} else {
-		row[10].SetInt64(0)
+			row[11].SetString(strings.ToUpper(hisHotRegion.HotRegionType), mysql.DefaultCollationName)
+			if hisHotRegion.HotDegree != 0 {
+				row[12].SetInt64(hisHotRegion.HotDegree)
+			} else {
+				row[12].SetNull()
```
Comment (nolouch) [inline executor/memtable_reader.go:980]: Why set null? can we directly set to 0?

File: store/helper/helper.go [hunk 2]
```diff
@@ -697,18 +713,25 @@ func (h *Helper) GetRegionsTableInfo(regionsInfo *RegionsInfo, schemas []*model.
 			}
 		}
 	}
+	sort.Sort(byTableStartKey(tables))
+	return tables
+}
 
-	if len(tables) == 0 || len(regions) == 0 {
+// ParseRegionsTableInfos parse the tables or indices in regions according to key range.
```
Comment (nolouch) [inline store/helper/helper.go:720]:
```suggestion
// ParseRegionsTableInfos parses the tables or indices in regions according to key range.
```

File: store/helper/helper.go [hunk 3]
```diff
@@ -670,19 +670,35 @@ func newPartitionTableWithKeyRange(db *model.DBInfo, table *model.TableInfo, par
 	}
 }
 
+// FilterMemDBs filter memory databases in the input schemas.
```
Comment (nolouch) [inline store/helper/helper.go:673]:
```suggestion
// FilterMemDBs filters memory databases in the input schemas.
```

**Review by nolouch — COMMENTED** [review]
rest LGTM

File: executor/memtable_reader.go (hunk 1 above)
Comment (IcePigZDB) [reply executor/memtable_reader.go]: Aggree, change to set it's origin value.

**Review by nolouch — APPROVED** [review]

File: executor/memtable_reader.go [hunk 4]
```diff
@@ -893,8 +891,8 @@ func (e *hotRegionsHistoryRetriver) retrieve(ctx context.Context, sctx sessionct
 	}
 	// Merge the results
 	var finalRows [][]types.Datum
-	allSchemas := sctx.GetInfoSchema().(infoschema.InfoSchema).AllSchemas()
 	tz := sctx.GetSessionVars().Location()
+	allSchemas := sctx.GetInfoSchema().(infoschema.InfoSchema).AllSchemas()
```
Comment (rleungx) [inline executor/memtable_reader.go:895]: Why change it, how about moving lines `894` and `895` before line `904`?

File: executor/memtable_reader_test.go [hunk 5]
```diff
@@ -1068,6 +1071,21 @@ func (s *testHotRegionsHistoryTableSuite) TearDownSuite(c *C) {
 	}
 	s.testInfoschemaTableSuiteBase.TearDownSuite(c)
 }
+func bytesKeyToHex(key []byte) string {
```
Comment (rleungx) [inline executor/memtable_reader_test.go:1074]: We already have this function in `helper.go`

File: executor/memtable_reader_test.go [hunk 6]
```diff
@@ -1068,6 +1071,21 @@ func (s *testHotRegionsHistoryTableSuite) TearDownSuite(c *C) {
...
+	return strings.ToUpper(hex.EncodeToString(key))
+}
+func (s *testHotRegionsHistoryTableSuite) newTableKeyRange(tableID int64) []string {
```
Comment (rleungx) [inline executor/memtable_reader_test.go:1077]: Please add an empty line.

File: executor/memtable_reader_test.go [hunk 7]
```diff
@@ -1076,113 +1094,141 @@ func (s *testHotRegionsHistoryTableSuite) TestTiDBHotRegionsHistory(c *C) {
...
-		// mysql table_id = 21, record_id = 1, table_name = STATS_META
+		// mysql table_id = 21, table_name = STATS_META
 		{"2019-10-10 10:10:13", "MYSQL", "STATS_META", "21", "<nil>", "<nil>", "3", "3", "33333", "0", "1", "READ", "99", "99", "99", "99"},
 		{"2019-10-10 10:10:14", "MYSQL", "STATS_META", "21", "<nil>", "<nil>", "4", "4", "44444", "0", "0", "WRITE", "99", "99", "99", "99"},
-		//      table_id = 131, record_id=1, deleted schema
-		{"2019-10-10 10:10:15", "UNKONW", "UNKONW", "131", "UNKONW", "<nil>", "5", "5", "55555", "0", "1", "READ", "99", "99", "99", "99"},
-		{"2019-10-10 10:10:16", "UNKONW", "UNKONW", "131", "UNKONW", "<nil>", "6", "6", "66666", "0", "0", "WRITE", "99", "99", "99", "99"},
-		// mysql table_id = 11, index_id = 1, index_value = 1, table_name = TABLES_PRIV, index_name = PRIMARY
+		// table_id = 1313, deleted schema
+		{"2019-10-10 10:10:15", "UNKONW", "UNKONW", "1313", "UNKONW", "<nil>", "5", "5", "55555", "0", "1", "READ", "99", "99", "99", "99"},
```
Comment (rleungx) [inline executor/memtable_reader_test.go:1104]: UNKONW -> UNKNOWN

File: executor/memtable_reader.go (hunk 4 above)
Comment (IcePigZDB) [reply executor/memtable_reader.go]: Done.

File: executor/memtable_reader_test.go (hunk 5 above)
Comment (IcePigZDB) [reply executor/memtable_reader_test.go]: I add an exported `BytesKeyToHex`.

File: executor/memtable_reader_test.go (hunk 6 above)
Comment (IcePigZDB) [reply executor/memtable_reader_test.go:1074]: Done.

File: executor/memtable_reader_test.go (hunk 7 above)
Comment (IcePigZDB) [reply executor/memtable_reader_test.go]: Done.

File: executor/memtable_reader_test.go [hunk 8]
```diff
@@ -1069,120 +1071,161 @@ func (s *testHotRegionsHistoryTableSuite) TearDownSuite(c *C) {
 	s.testInfoschemaTableSuiteBase.TearDownSuite(c)
 }
 
+func (s *testHotRegionsHistoryTableSuite) newTableKeyRange(tableID int64) []string {
+	sk, ek := tablecodec.GetTableHandleKeyRange(tableID)
+	startKey := helper.BytesKeyToHex(codec.EncodeBytes(nil, sk))
+	endKey := helper.BytesKeyToHex(codec.EncodeBytes(nil, ek))
+	return []string{startKey, endKey}
+}
+func (s *testHotRegionsHistoryTableSuite) newIndexKeyRange(tableID, indexID int64) []string {
```
Comment (rleungx) [inline executor/memtable_reader_test.go:1080]: ditto

**Review by rleungx — APPROVED** [review]
The rest LGTM

File: executor/memtable_reader.go [hunk 9]
```diff
@@ -903,14 +899,18 @@ func (e *hotRegionsHistoryRetriver) retrieve(ctx context.Context, sctx sessionct
 		Store:       tikvStore,
 		RegionCache: tikvStore.GetRegionCache(),
 	}
+	tz := sctx.GetSessionVars().Location()
+	allSchemas := sctx.GetInfoSchema().(infoschema.InfoSchema).AllSchemas()
```
Comment (JmPotato) [inline executor/memtable_reader.go:903]: Should we use `ok` to assert the `infoschema.InfoSchema` first to prevent potential panic?

**Comment by rleungx** [comment]
BTW, I find the results of the key rate in your description is always 0.

File: executor/memtable_reader.go (hunk 9 above)
Comment (IcePigZDB) [reply executor/memtable_reader.go]: Use `GetTxnInfoSchema` func avoid this convert, #30574.
```go
	allSchemas := sessiontxn.GetTxnManager(sctx).GetTxnInfoSchema().AllSchemas()
```

Comment (IcePigZDB) [reply executor/memtable_reader.go]:
```go
	// Deprecated: Use TxnManager.GetTxnInfoSchema to get the current schema in session
	GetInfoSchema() InfoschemaMetaVersion
```

File: executor/memtable_reader_test.go (hunk 8 above)
Comment (IcePigZDB) [reply executor/memtable_reader_test.go]: Done by export `NewIndexWithKeyRange` and `NewTableWithKeyRange` for test only now.

**Comment by IcePigZDB** [comment]
check_dev_2

> BTW, I find the results of the key rate in your description is always 0.

It's record on PD side, I will try to dig it on another day.

**Comment by IcePigZDB** [comment]
> BTW, I find the results of the key rate in your description is always 0.

Hot regions through [sysbench](https://docs.pingcap.com/tidb/stable/benchmark-tidb-using-sysbench/) with small table size 100,000 rather than 1000,000.
```bash
sysbench --config-file=sb_config oltp_point_select --tables=32 --table-size=100000 prepare

sysbench --config-file=sb_config oltp_point_select --tables=32 --table-size=100000 run
sysbench --config-file=sb_config oltp_update_index --tables=32 --table-size=100000 run
sysbench --config-file=sb_config oltp_read_only --tables=32 --table-size=100000 run
```
And log the result got from 
`server/handler.go/PackHistoryHotReadRegions`  ->
`server/handler.go/GetHotReadRegions`  :
```go
// GetHotReadRegions gets all hot read regions stats.
func (h *Handler) GetHotReadRegions() *statistics.StoreHotPeersInfos {
	c, err := h.GetRaftCluster()
	if err != nil {
		return nil
	}
	return c.GetHotReadRegions()
}
```
I put the log [here](https://github.com/IcePigZDB/discussion-ref/blob/master/pd_stat_keyrate/keyrate.log), `keyrate` result is always 0 ` KeyRate: (float64) 0,`. 
We check the get code on PD and TiDB, We think it's right, if the hot method is right. Since the `queryRate` and `flowBytes` data are normal, I think there is a deeper reason in `hotScheduler`/ `HotCache`.

File: executor/memtable_reader.go (hunk 9 above)
Comment (nolouch) [reply executor/memtable_reader.go]: cc @JmPotato

**Pull request merged**